### DISCUSSION TITLE
Should we remove test types

### DISCUSSION THREAD

**Discussion by sksamuel** [kotest/kotest#4055] [6.0 changes]
Builds on https://github.com/kotest/kotest/pull/4053 and #4054 

By not needing the test type enum in the TC listener, we have the possibility of removing test type entirely.
We would have to remove IsolationMode.InstancePerLeaf (since we don't know in advance if something is a leaf or not).

But it would mean we can remove StringSpec (as FreeSpec would do the same).
And it would mean we could arbitrarily nest property tests and data tests.

Other frameworks keep the hard coded distinction because they don't allow tests to be dynamically generated at runtime like Kotest does.

**Comment by sschuberth** [comment]
> But it would mean we can remove StringSpec (as FreeSpec would do the same).

Speaking for myself and / or the [OSS Review Toolkit](https://github.com/oss-review-toolkit/ort) project, where we use `StringSpec` a lot, but `FreeSpec` in only a single place where we need more than two levels of nesting:

What I don't like about `FreeSpec` is that you have to use `-` as a marker. If that could be the `should`-keyword as well, then obviously `FreeSpec` would just be a generalization of `StringSpec`, and personally I'd have no problem with dropping the latter (or making `StringSpec` an typealias for `FreeSpec`).

**Reply by Kantis** [reply]
Being able to nest stuff in StringSpec without the `-` marker would be awesome!

**Reply by sschuberth** [reply]
> What I don't like about `FreeSpec` is that you have to use `-` as a marker.

I guess I only now fully understand @sksamuel's "But it would mean we can remove StringSpec (as FreeSpec would do the same)" from [above](https://github.com/kotest/kotest/discussions/4055#discussion-6766070):

The `-` syntax is only needed to add test containers, i.e. it's not needed for `FreeSpec`s that only have leaf tests, which case `StringSpec` syntax exactly equals `FreeSpecs`.

So indeed it would sense to me drop any separate `StringSpec` implementation then, in favor of making `StringSpec` a `typealias` for `FreeSpec`.

**Comment by olegivo** [comment]
We have a thousands of tests in several projects that rely on `IsolationMode.InstancePerLeaf` and KoinExtension. It is very handy to combine behaviors on several levels of the test without needing to clear up the state of Koin container after every sibling test, so these tests can be run isolated from each other in case of `InstancePerLeaf`.  
Please, don't break this behavior

**Comment by sksamuel** [comment]
Closing, as we will be sticking with the current behavior based on feedback.

**Reply by olegivo** [reply]
There was critical breaking feedback

**Reply by sksamuel** [reply]
Yep, nothing is changing
